Import schema from enterprise-chef-server-schema

stevendanna opened the pull request:
Enterprise-chef-server-schema often needs to change alongside the code
that makes use of the new schema.  Having the schema in a separate
repository introduced unncessary friction into the development
process.

deps/chef-server-schema is an import of chef-server-schema at 1.0.4

If you have stumbled upon this repository via git-log or git-blame and
need more history, see the following upstream repositories:

https://github.com/chef/enterprise-chef-server-schema

and

https://github.com/chef/chef-server-schema

stevendanna commented:
See also: https://github.com/chef/opscode-omnibus/pull/711

Todo/Open Questions:

- [ ] Determine why tests are failing on travis but working locally
- [ ] What should the next "sqitch tag" be?  
- [ ] How would this impact Hosted Chef?

sdelano commented:
> What should the next `sqitch tag` be?

> How would this impact Hosted Chef?

When we end up moving the oc_erchef tarball package deployment into Travis, we could also package up the schema for the deploy to pull down as well.

tylercloke commented on schema/Makefile:
```diff
@@ -0,0 +1,93 @@
+# The version of the open source schema that this version of the
+# enterprise schema depends on.  Change this as new versions become
+# available.
+OSC_SCHEMA_VERSION=1.0.4
```
What happened to the chef-server-schema Makefile? Does this affect upgrades?

stevendanna replied:
@tylercloke In the other PR I updated all the upgrade logic to run sqitch directly.  I think we could probably nuke these make files completely and maybe just make a single schema target in the top level makefile if we want one for some reason.

tylercloke replied:
^ :+1: That would be great.

sdelano commented:
A follow-on question here:

It looks like we've also consolidated `chef-server-schema` into this repo as well. Is there a path forward where we can move this out of the `deps` subdir and expose it at the top level?

stevendanna commented:
> It looks like we've also consolidated chef-server-schema into this repo as well. Is there a path forward where we can move this out of the deps subdir and expose it at the top level?

We could do that now.  We'll just need to change the path in omnibus and in the test helpers.

tylercloke commented:
As long as this work with OCS and EC 11.latest upgrade to this and we answer the remaining questions, I am a huge :+1:

stevendanna commented:
> Did you get a chance to put more thought into this? It sounds like the options are:
> Bump the oc_erchef tag to be above the schema tag and move them forward in step
> Just use text-based tags in the schema

Those are definitely our options.  I don't have a strong leaning one way or the other.  I think we could just roll forward with the versioning in the plan file and accept that there won't be a git tag to go with them for now.  Since we aren't using the Makefile for deploys I don't think it matters on the technical side.

tylercloke commented:
I would be a fan of the schema version matching the oc_erchef version just for simplicity.

marcparadise commented on schema/Makefile:
```diff
@@ -0,0 +1,93 @@
+# The version of the open source schema that this version of the
+# enterprise schema depends on.  Change this as new versions become
+# available.
```
I'd change this to say 'No further updates will be made to this schema'.  In fact, I wouldn't object to using the number directly without making it a variable, because it won't be changed.

marcparadise commented:
So can we have sqitch use a different filename other than sqitch.plan? If so, we might be able to keep all of the sql in t/, deploy/ and revert/? 

Alternatively, would it be possible to just combine the plan files? (I suspect not, because of the hashing...)

marcparadise commented:
:+1: other than minor comments/questions above.

stevendanna commented:
@chef/lob I've updated this and the related opscode-omnibus branch to use schema/baseline rather than schema/deps/chef-server-schema, updated the Makefiles and READMEs to hopefully prevent confusion, and made minor updates to the test and deploy logic to account for the directory move.

I've tested the following (on a Standalone):

- Clean install
- Upgrade from 12.0.0
- Upgrade from EC 11.2.6

I can't see why this would behave different on Tier or HA.  Open Source 11 -> 12 upgrades take a completely different path, so I think we should be OK there as well.

From the discussion, I think that the next "sqitch tag" should line up with the Chef Server (no oc_erchef) version we plan to ship the change in.  While it might be a little confusing, at least it will make it clear where a given change shipped.  This might also help in the future if we combine opscode-omnibus and oc_erchef.

For Hosted, the schema is part of the release that the oc_erchef project creates, so it is likely we can construct some suitable deploy process using that.

Any other concerns that can't be taken care of iteratively that people can think of?  If not, I'd vote to merge this so new feature work can build on top of it.

marcparadise commented:
Honestly given that the schema doesn't use the version numbers as aanything other than tags, I think we should drop it and start sqitch tagging by feature name.  That'll make it easier to resolve conflicts as well, without having to edit version numbers already in the plan file. 

MWR.

stevendanna commented:
@marcparadise Ahh, interesting idea.
